File: src/org/jbali/jmsrpc/TextMessageService.java
package org.jbali.jmsrpc;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.Arrays;
import java.util.NoSuchElementException;

import org.jbali.json.JSONArray;
import org.jbali.reflect.Methods;
import org.jbali.serialize.JavaJsonSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;

public class TextMessageService {
	
	public static final int RQIDX_METHOD = 0;
	
	public static final int RSIDX_STATUS = 0;
	public static final int RSIDX_RESPONSE = 1;
	
	public static final int STATUS_OK = 1;
	public static final int STATUS_ERROR = 0;
	
	
	
	private static final Logger log = LoggerFactory.getLogger(TextMessageService.class);
	
	private final Object endpoint;
	private final ImmutableMap<String, Method> methods;
	
	public TextMessageService(Object endpoint) {
		Preconditions.checkNotNull(endpoint);
		this.endpoint = endpoint;
		methods = Methods.mapPublicMethodsByName(endpoint.getClass());
	}
	
	public String handleRequest(String request) {
		
		JSONArray response;
		String methName = "?";
		boolean requestLogged = false;
		
		try {
			
			// parse request json
			JSONArray reqJson;
			try {
				reqJson = new JSONArray(request);
			} catch (Throwable e) {
				throw new IllegalArgumentException("Could not parse request", e);
			}
			
			// determine method
			methName = reqJson.getString(RQIDX_METHOD);
			Method method = methods.get(methName);
			if (method == null) throw new NoSuchElementException("Unknown method " + methName);
			
			// read arguments
			Parameter[] pars = method.getParameters();
			Object[] args = new Object[pars.length];
			for (int p = 0; p < pars.length; p++) {
				int ri = p+1;
				Object arg;
				if (reqJson.length() < ri+1) {
					// this parameter has no argument
					if (!requestLogged) {
						log.info(methName + ":");
						requestLogged = true;
					}
					log.info("- Arg #" + p + " omitted");
//					arg = pars[p].getType() == Maybe.class ? Maybe.unknown() : null;
					arg = null; // let's hope that's sufficient
				} else {
					arg = JavaJsonSerializer.unserialize(reqJson.get(ri));
				}
				args[p] = arg;
			}
			// check for more args than used
			int extraArgs = reqJson.length() - 1 - pars.length;
			if (extraArgs > 0) {
				if (!requestLogged) {
					log.info(methName + ":");
					requestLogged = true;
				}
				log.info("- " + extraArgs + " args too many ignored.");
			}
			
			// execute
			Object ret;
			try {
				ret = method.invoke(endpoint, args);
			} catch (InvocationTargetException | ExceptionInInitializerError e) {
				// InvocationTargetException: actual exception inside method
				// ExceptionInInitializerError: always unchecked (initializers can't throw checked)
				throw e.getCause();
			} catch (IllegalAccessException | NullPointerException e) {
				// should not happen
				throw new AssertionError(e);
			}
			// TODO return more info for mismatched arguments, such as index
			
			// return response
			response = JSONArray.create(STATUS_OK, JavaJsonSerializer.serialize(ret));
			
		} catch (Throwable e) {
			
			log.warn("Error in text request for method " + methName, e);
			
			// remove the current stack trace from the error trace,
			// because it's not relevant to the client.
			StackTraceElement[] orgTrace = e.getStackTrace();
			StackTraceElement[] locTrace = new Throwable().getStackTrace();
			e.setStackTrace(Arrays.copyOfRange(orgTrace, 0, orgTrace.length - locTrace.length));
			
			try {
				response = JSONArray.create(STATUS_ERROR, JavaJsonSerializer.serialize(e));
			} catch (Throwable serEr) {
				log.warn("Error while serializing error", serEr);
				response = JSONArray.create(STATUS_ERROR, JavaJsonSerializer.serialize(new RuntimeException("Error occurred but could not be serialized")));
			} finally {
				e.setStackTrace(orgTrace);
			}
		}
		
		try {
			return response.toString(2);
		} catch (Throwable e) {
			log.warn("Error toStringing JSON response", e);
			return "[0, null]";
		}
		
	}
	
}

File: src/main/java/org/jbali/jmsrpc/TextMessageService.kt
package org.jbali.jmsrpc

import org.jbali.errors.removeCurrentStack
import org.jbali.json.JSONArray
import org.jbali.reflect.Methods
import org.jbali.serialize.JavaJsonSerializer
import org.jbali.util.onceFunction
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

private val log = LoggerFactory.getLogger(TextMessageService::class.java)!!

// TODO simple kotlin-serialization JSON instead of JavaJsonSerializer.
//      perhaps migrate to that slowly with an annotation per method.
class TextMessageService<T : Any>(
        private val iface: Class<out T>,
        private val svcName: String = iface.name,
        private val endpoint: T
) {

    private val methods: Map<String, Method> =
            Methods.mapPublicMethodsByName(iface)
                .mapKeys { it.key.toLowerCase() }

    fun handleRequest(request: String): String {

        var methName = "?"
        val logTheRequest = onceFunction { log.info("In text request $svcName.$methName:") }

        val response = try {

            // parse request json
            val reqJson = try {
                JSONArray(request)
            } catch (e: Throwable) {
                throw IllegalArgumentException("Could not parse request", e)
            }

            // determine method
            methName = reqJson.getString(RQIDX_METHOD)
            val method = methods[methName.toLowerCase()] ?: throw NoSuchElementException("Unknown method '$methName'")

            // read arguments
            val pars = method.parameters

            val args = method.parameters.mapIndexed { p, par ->
                val indexInReq = p + 1
                if (reqJson.length() < indexInReq + 1) {
                    // this parameter has no argument
                    logTheRequest()
                    log.info("- Arg #" + p + " (" + par.type + " " + par.name + ") omitted")
                    null // let's hope that's sufficient
                } else {
                    JavaJsonSerializer.unserialize(reqJson.get(indexInReq))
                }
            }.toTypedArray()

            // check for more args than used
            val extraArgs = reqJson.length() - 1 - pars.size
            if (extraArgs > 0) {
                logTheRequest()
                log.info("- $extraArgs args too many ignored.")
            }

            // execute
            val ret = try {
                method.invoke(endpoint, *args)
            } catch (e: InvocationTargetException) {
                // InvocationTargetException: actual exception inside method.
                throw e.cause!!
            } catch (e: ExceptionInInitializerError) {
                // ExceptionInInitializerError: always unchecked (initializers can't throw checked).
                throw e.cause!!
            } catch (e: IllegalAccessException) {
                // IllegalAccessException: method is public, should not happen.
                throw RuntimeException("TextMessageService internal error", e)
            } catch (e: NullPointerException) {
                // NullPointerException: endpoint is not null, should not happen.
                throw RuntimeException("TextMessageService internal error", e)
            }

            // return response
            JSONArray.create(STATUS_OK, JavaJsonSerializer.serialize(ret))!!

        } catch (e: Throwable) {

            // remove the current stack trace from the error stacktraces,
            // because it's not relevant to the client or logs.
            // TODO remove up to the endpoint as well, e.g.:
            //			at com.blabla.DataServer$RemoteDataServerImpl.authByXId(DataServer.java:123) ~[dataserver.jar:na]
            //			...
            //			at org.jbali.jmsrpc.TextMessageService.handleRequest(TextMessageService.java:95) ~[bali.jar:na]
            e.removeCurrentStack()

            logTheRequest()
            log.warn("Error handling request", e)

            try {
                JSONArray.create(STATUS_ERROR, JavaJsonSerializer.serialize(e))!!
            } catch (serEr: Throwable) {
                serEr.removeCurrentStack()
                log.warn("!! Error while serializing error", serEr)
                JSONArray.create(STATUS_ERROR, JavaJsonSerializer.serialize(RuntimeException("Error occurred but could not be serialized (see log for details)")))!!
            }

        }

        return try {
            response.toString(2)
        } catch (e: Throwable) {
            log.warn("Error toStringing JSON response", e)
            "[0, null]"
        }

    }

    companion object {

        const val RQIDX_METHOD = 0

        const val RSIDX_STATUS = 0
        const val RSIDX_RESPONSE = 1

        const val STATUS_OK = 1
        const val STATUS_ERROR = 0

    }

}
